Introduce environment.profileRelativeSessionVariables

worldofpeace opened the pull request:
###### Motivation for this change
Fixes the general issue found in #68383. This has been useful for every display-manager.

Thanks @rycee for https://github.com/NixOS/nixpkgs/commit/573648903b364fa5c322e3890cfd209bbc9a09da it was magically just the commit I needed.
Originates from #67942.

This will allow me to fix https://github.com/elementary/greeter/issues/318 and hopefully update it to its latest release.

###### Things to test for display-managers

- [x] lightdm greeters have proper theming

- [x] correct cursor theme is used in gdm

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

rycee committed:
nixos/system-environment: introduce environment.profileRelativeSessionVariables

There is a need for having sessionVariables set relative to the Nix Profiles.
Such as in #68383.

worldofpeace committed:
nixos/terminfo: use profileRelativeSessionVariables

worldofpeace committed:
nixos/xdg/icons: use profileRelativeSessionVariables

worldofpeace committed:
nixos/gdm: don't set XCURSOR_PATH

worldofpeace committed:
nixos/sddm: don't set XDG_DATA_DIRS

environment.profileRelativeSessionVariables should make this unneeded.

worldofpeace committed:
nixos/enso-os: don't wrap in module

worldofpeace committed:
nixos/lightdm-gtk-greeter: don't wrap in module

worldofpeace committed:
lightdm-mini-greeter: use wrapGAppsHook

worldofpeace committed:
lightdm-mini-greeter: put xgreeters in passthru

rycee commented:
I'm glad the changes in #67942 are more generally useful :+1:

Originally I didn't have a `profileRelativeSessionVariables` option and instead simply put `profileRelativeEnvVars` in the pam configuration with `$HOME` and `$USER` replaced by `@{HOME}` and `@{PAM_USER}`. Not sure it if it perhaps would be good to return to that idea. I do kind of like separating the PAM and the shell variables, though since it would allow people to use the pam variable syntax.

worldofpeace commented:
> Originally I didn't have a profileRelativeSessionVariables option and instead simply put profileRelativeEnvVars in the pam configuration with $HOME and $USER replaced by @{HOME} and @{PAM_USER}. Not sure it if it perhaps would be good to return to that idea. I do kind of like separating the PAM and the shell variables, though since it would allow people to use the pam variable syntax.

I actually like it this way, I think I'd expect `profileRelativeEnvVars` to just be in the shell.
A further idea here is to use `sessionVariables` everywhere :smile:

worldofpeace committed:
nixos/environment: set GTK_DATA_PREFIX

Many desktop environment modules are already setting
this so it already makes sense to just do this globally.

worldofpeace commented:
We currently have it setup for `environment.variables` to always have whatever is set in `config.environment.sessionVariables`.
https://github.com/NixOS/nixpkgs/blob/f595f28d5563284b3418ff83e0f8d3fe5b88472e/nixos/modules/config/shells-environment.nix#L158

Should we do the same for `environment.profileRelativeEnvVars` and only set `environment.profileRelativeSessionVariables` everywhere?

worldofpeace committed:
nixos/environment: set GTK_EXE_PREFIX

michaelpj commented on nixos/modules/config/shells-environment.nix:
```diff
@@ -157,6 +157,8 @@ in
     # terminal instead of logging out of X11).
     environment.variables = config.environment.sessionVariables;
 
+    environment.profileRelativeEnvVars = config.environment.profileRelativeSessionVariables;
```
I thought this seemed unnecesssary, then I read the comment on the assignment above.  Maybe move this up a line so it obviously belongs with the previous one.

michaelpj commented on nixos/modules/programs/environment.nix:
```diff
@@ -30,7 +32,7 @@ in
 
     # TODO: move most of these elsewhere
-    environment.profileRelativeEnvVars =
+    environment.profileRelativeSessionVariables =
```
Should we be using `environment.sessionVariables` for the assignments to `environment.variables` above as well?

worldofpeace replied:
I'm going to migrate all uses of `environment.variables` to `sessionVariables` where appropriate (in another PR).
Also, I think that would infinitely recurse.

michaelpj commented on nixos/modules/config/system-environment.nix:
```diff
@@ -18,25 +23,76 @@ in
+    environment.profileRelativeSessionVariables = mkOption {
+      type = types.attrsOf (types.listOf types.str);
+      example = { PATH = [ "/bin" ]; MANPATH = [ "/man" "/share/man" ]; };
+      description = ''
```
It would be great to have some guidance about when to use this versus `environment.profileRelativeEnvVars` (likewise `sessionVariables` vs `environment.variables`). I'm not sure where that should live, though.

michaelpj commented on nixos/modules/services/x11/display-managers/gdm.nix:
```diff
@@ -142,8 +142,6 @@ in
           GDM_X_SERVER_EXTRA_ARGS = toString
             (filter (arg: arg != "-terminate") cfg.xserverArgs);
           XDG_DATA_DIRS = "${cfg.session.desktops}/share/";
```
This line also seems like it should be redundant now?

worldofpeace replied:
I don't think `session.desktops` is going to be installed/linked.

michaelpj commented on nixos/modules/services/x11/display-managers/sddm.nix:
```diff
@@ -219,8 +219,6 @@ in
         # Load themes from system environment
         QT_PLUGIN_PATH = "/run/current-system/sw/" + pkgs.qt5.qtbase.qtPluginPrefix;
         QML2_IMPORT_PATH = "/run/current-system/sw/" + pkgs.qt5.qtbase.qtQmlPrefix;
```
These two seem like they should also be handled. `QT_PLUGIN_PATH` is set, but to something different. `QML2_IMPORT_PATH` isn't set. Not sure if this is suspicious or not.

worldofpeace replied:
I'm not familiar with these things for QT, so I choose to not change this so the PR doesn't get blocked. Looking at git blame it was intentional a while ago, so I'd have to check with @ttuegel.

ttuegel replied:
`QT_PLUGIN_PATH` and `QML2_IMPORT_PATH` should not be set system-wide because the paths have version-specific prefixes. For historical reasons, they actually are set system-wide at the moment, but it is important that SDDM keep these locally-hardcoded paths.

worldofpeace replied:
Thanks, good that I didn't try to introduce a change to remove this.

michaelpj reviewed (approved):
This looks great!

Does this activate early enough that systemd system services will get it? I'm sure I've had problems before due to them not having `XDG_DATA_DIRS` set properly.

worldofpeace commented:
> Does this activate early enough that systemd system services will get it? I'm sure I've had problems before due to them not having XDG_DATA_DIRS set properly.

I did `systemctl --user show-environment` and it has the profile relative environment variables.

michaelpj commented:
> I did systemctl --user show-environment and it has the profile relative environment variables.

I guess it's too much to hope that it'll get into `systemctl show-environment` for the *system* services too. I guess those inherit from systemd itself which probably starts earlier.
